crates/driver/src/domain/competition/solution/encoding.rs

```diff
@@ -7,3 +22,330 @@ pub enum Strategy {
+            }
+        }
+    }
+
+    // Encode allowances
+    for approval in approvals {
+        interactions.push(approve(&approval.0))
+    }
+
+    // Encode interaction
+    for interaction in solution.interactions() {
+        if matches!(internalization, settlement::Internalization::Enable)
+            && interaction.internalize()
+        {
+            continue;
+        }
+
+        interactions.push(match interaction {
+            competition::solution::Interaction::Custom(interaction) => eth::Interaction {
+                value: interaction.value.into(),
+                target: interaction.target.0.into(),
+                call_data: interaction.call_data.clone(),
+            },
+            competition::solution::Interaction::Liquidity(interaction) => {
+                // Todo account for slippage
+                let input = MaxInput(interaction.input);
+                let output = ExactOutput(interaction.output);
+
+                match interaction.liquidity.kind.clone() {
+                    liquidity::Kind::UniswapV2(pool) => {
+                        pool.swap(&input, &output, &contract.address().into()).ok()
+                    }
+                    liquidity::Kind::UniswapV3(pool) => {
+                        pool.swap(&input, &output, &contract.address().into()).ok()
+                    }
+                    liquidity::Kind::BalancerV2Stable(pool) => {
+                        pool.swap(&input, &output, &contract.address().into()).ok()
+                    }
+                    liquidity::Kind::BalancerV2Weighted(pool) => {
+                        pool.swap(&input, &output, &contract.address().into()).ok()
+                    }
+                    liquidity::Kind::Swapr(pool) => {
+                        pool.swap(&input, &output, &contract.address().into()).ok()
+                    }
+                    liquidity::Kind::ZeroEx(limit_order) => limit_order.to_interaction(&input).ok(),
+                }
+                .ok_or(Error::InvalidInteractionExecution(interaction.clone()))?
+            }
+        })
+    }
+
+    let tx = contract
+        .settle(
+            tokens.clone(),
+            clearing_prices.clone(),
+            trades.iter().map(codec::trade).collect(),
+            [
+                pre_interactions.iter().map(codec::interaction).collect(),
+                interactions.iter().map(codec::interaction).collect(),
+                post_interactions.iter().map(codec::interaction).collect(),
+            ],
+        )
+        .into_inner();
+
+    // Encode the auction id into the calldata
+    let mut calldata = tx.data.unwrap().0;
+    calldata.extend(auction_id.to_be_bytes());
+
+    Ok(eth::Tx {
+        from: solution.solver().address(),
+        to: contract.address().into(),
+        input: calldata.into(),
+        value: Ether(0.into()),
+        access_list: Default::default(),
+    })
+}
+
+fn order_flags(order: &competition::Order) -> eth::U256 {
```

Comment:
We could use the crate `bitflags` to handle these operations: https://docs.rs/bitflags/latest/bitflags/

Reply:
True, however I copied the working code from: https://github.com/cowprotocol/services/blob/81831022cc41a05e76547b90e87694dc349a0388/crates/shared/src/encoded_settlement.rs#L49

So I'm quite sure this is correct.